app/imaging/decode.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package imaging
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"io"
+	"sync"
+
+	_ "github.com/oov/psd"
+	_ "golang.org/x/image/bmp"
+	_ "golang.org/x/image/tiff"
+)
+
+// DecoderOpts holds configuration options for an image decoder.
+type DecoderOpts struct {
+	// The level of concurrency for the decoder. This defines a limit on the
+	// number of concurrently running encoding goroutines.
+	ConcurrencyLevel int
+}
+
+func (o *DecoderOpts) Validate() error {
+	if o.ConcurrencyLevel < 0 {
+		return errors.New("ConcurrencyLevel must be non-negative")
+	}
+	return nil
+}
+
+// Decoder holds the necessary state to decode images.
+// This is safe to be used from multiple goroutines.
+type Decoder struct {
+	sem  chan struct{}
+	opts DecoderOpts
+}
+
+// NewDecoder creates and returns a new image decoder with the given options.
+func NewDecoder(opts DecoderOpts) (*Decoder, error) {
+	var d Decoder
+	if err := opts.Validate(); err != nil {
+		return nil, fmt.Errorf("imaging: error validating decoder options: %w", err)
+	}
+	if opts.ConcurrencyLevel > 0 {
+		d.sem = make(chan struct{}, opts.ConcurrencyLevel)
+	}
+	d.opts = opts
+	return &d, nil
+}
+
+// Decode decodes the given encoded data and returns the decoded image.
+func (d *Decoder) Decode(rd io.Reader) (img image.Image, format string, err error) {
+	if d.opts.ConcurrencyLevel != 0 {
+		d.sem <- struct{}{}
+		defer func() { <-d.sem }()
+	}
+
+	img, format, err = image.Decode(rd)
+	if err != nil {
+		return nil, "", fmt.Errorf("imaging: failed to decode image: %w", err)
+	}
+
+	return img, format, nil
+}
+
+// DecodeMemBounded works similarly to Decode but also returns a release function that
+// must be called when access to the raw image is not needed anymore.
+// This sets the raw image data pointer to nil in an attempt to help the GC to re-use the underlying data as soon as possible.
+func (d *Decoder) DecodeMemBounded(rd io.Reader) (img image.Image, format string, releaseFunc func(), err error) {
```

Comment:
I am wondering if instead of a release function, doing an `runtime.GC` makes any difference? Or is the runtime unable to reclaim the memory until you explicitly mark them as nil?

Reply:
I am not sure I wanted to go as far as forcing the GC call which may have other performance implications in itself and doesn't look like something recommended. The idea here was just to do our best to help with re-using memory by setting its reference to nil. I believe GC behaviour is overall non-deterministic so that's not a guarantee of memory release but it will help since the caller might be doing a lot more before the image reference goes out of scope.

Reply:
Agreed, ideally we'd like something like a shared buffer pool to have more control on this front but we need some help from the API side to achieve that. Might be an idea for a possible upstream contribution who knows :p